timesketch/views/auth.py

```diff
@@ -134,3 +148,71 @@ def logout():
     """
     logout_user()
     return redirect(url_for(u'user_views.login'))
+
+
+@auth_views.route(u'/login/google_openid_connect/', methods=[u'GET'])
+def google_openid_connect():
+    """Handler for the Google OpenID Connect callback.
+
+    Reference:
+    https://developers.google.com/identity/protocols/OpenIDConnect
+
+    Returns:
+        Redirect response.
+    """
+    error = request.args.get(u'error', None)
+
+    if error:
+        current_app.logger.error(u'OAuth2 flow error: {}'.format(error))
+        return abort(HTTP_STATUS_CODE_BAD_REQUEST)
+
+    try:
+        code = request.args[u'code']
+        client_csrf_token = request.args.get(u'state')
+        server_csrf_token = session[CSRF_KEY]
+    except KeyError:
+        return abort(HTTP_STATUS_CODE_BAD_REQUEST)
+
+    if client_csrf_token != server_csrf_token:
+        return abort(HTTP_STATUS_CODE_BAD_REQUEST, u'Invalid CSRF token')
+
+    try:
+        encoded_jwt = get_encoded_jwt_over_https(code)
+    except JwtFetchError:
+        return abort(HTTP_STATUS_CODE_BAD_REQUEST)
+
+    try:
+        discovery_document = get_oauth2_discovery_document()
+    except DiscoveryDocumentError:
+        return abort(HTTP_STATUS_CODE_BAD_REQUEST)
+
+    algorithm = discovery_document[u'id_token_signing_alg_values_supported'][0]
+    expected_audience = current_app.config.get(u'GOOGLE_OIDC_CLIENT_ID')
+    expected_domain = current_app.config.get(u'GOOGLE_OIDC_HOSTED_DOMAIN')
+    expected_issuer = discovery_document[u'issuer']
+
+    # Fetch the public key and try to validate the JWT.
+    try:
+        public_key = get_public_key_for_jwt(
+            encoded_jwt, discovery_document[u'jwks_uri'])
+        validated_jwt = validate_jwt(
+            encoded_jwt, public_key, algorithm, expected_audience,
+            expected_issuer, expected_domain)
+    except (JwtValidationError, JwtKeyError) as e:
+        current_app.logger.error(u'{}'.format(e))
+        return abort(HTTP_STATUS_CODE_UNAUTHORIZED)
+
+    validated_email = validated_jwt.get(u'email')
+    user_whitelist = current_app.config.get(u'GOOGLE_OIDC_USER_WHITELIST')
+
+    # Check if the authenticating user is on the whitelist.
+    if user_whitelist:
+        if validated_email not in user_whitelist:
+            return abort(HTTP_STATUS_CODE_UNAUTHORIZED)
+
+    user = User.get_or_create(username=validated_email, name=validated_email)
```

Comment:
One of `username` or `name` should be enough to have a unique user, right?

Reply:
True, but the way the helper works I need both here so it will be created correctly. Might change this later, but this PR is not the place :)

Reply:
SG